std/assembly/util/uri.ts

```diff
@@ -0,0 +1,260 @@
+import { E_URI_MALFORMED } from "./error";
+import { CharCode } from "./string";
+
+// @ts-ignore: decorator
+@lazy export const URI_UNSAFE = memory.data<u8>([
+/* skip 32 + 1 always set to '1' head slots
+  */ 0, 1, 0, 0, 1, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0,
+  0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 1, 0, 1, 0,
+  0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0,
+  0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 1, 1, 1, 1, 0,
+  1, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0,
+  0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 1, 1, 1, 0, /*
+  skip 128 + 1 always set to '1' tail slots */
+]);
+
+// @ts-ignore: decorator
+@lazy export const URL_UNSAFE = memory.data<u8>([
+/* skip 32 + 1 always set to '1' head slots
+  */ 0, 1, 1, 1, 1, 1, 0, 0, 0, 0, 1, 1, 0, 0, 1,
+  0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 1, 1, 1, 1, 1, 1,
+  1, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0,
+  0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 1, 1, 1, 1, 0,
+  1, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0,
+  0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 1, 1, 1, 0, /*
+  skip 128 + 1 always set to '1' tail slots */
+]);
+
+// @ts-ignore: decorator
+@lazy export const URI_RESERVED = memory.data<u8>([
+  /*  skip 32 + 3 always set to '0' head slots
+        */ 1, 1, 0, 1, 0, 0, 0, 0, 1, 1, 0, 0, 1,
+  0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 1, 1, 0, 1, 0, 1,
+  1, /* skip 191 always set to '0' tail slots */
+]);
+
+export function encode(dst: usize, src: usize, len: usize, table: usize): usize {
+  var i: usize = 0, offset: usize = 0, outSize = len << 1;
+
+  while (i < len) {
+    let org = i;
+    let c: u32, c1: u32;
+    do {
+      c = <u32>load<u16>(src + (i << 1));
+      if (c - 33 < 94) { // 127 - 33
+        if (load<u8>(table + (c - 33))) break;
+      } else break;
+    } while (++i < len);
+
+    if (i > org) {
+      let size = i - org << 1;
+      if (offset + size > outSize) {
+        outSize = offset + size;
+        dst = __renew(dst, outSize);
+      }
+      if (size == 2) {
+        store<u16>(dst + offset, load<u16>(src + (org << 1)));
+      } else if (size == 4) {
+        store<u32>(dst + offset, load<u32>(src + (org << 1)));
+      } else {
```

Comment:
I don't have any information on the performance characteristics of `memory.copy` yet, but last I heard was that it is a CPU instruction. I'd just use `memory.copy` here, since if it ever turns out that it falls short for small values, we'd want to revisit all of stdlib for this pattern anyway.

Reply:
`memory.copy` may optimize to serial of loads and stores only for constant sizes like `memory.copy(a, b, 4)`. But if `size` is dynamic `memory.copy` for short lenghts will be significantly slower due to memory.copy have several bound checkers (if we talk about bulk memory feture on engines)

Reply:
Are you referring to Binaryen or more generally to the future instruction? In Binaryen that's true, yeah, but I'd assume that an instruction can do better?

Reply:
I refer to Binaryen (basically I'm optimize this path) as well as exec engines like wasmtime and v8

Reply:
Hmm, I have a feeling that we shouldn't try to over optimize this as long as we don't know for sure that the `memory.copy` instruction (only waiting for Safari at this point) doesn't perform well enough *on average* here, since if it does, we have two potentially unnecessary branches plus their code size. To know for sure, though, we'd have to benchmark this somehow, which I guess we can do in V8?